README.md
# TaxCloud

[![Build Status](https://github.com/txcrb/tax_cloud/actions/workflows/test.yml/badge.svg?branch=master)](https://github.com/txcrb/tax_cloud/actions)

[TaxCloud](http://www.taxcloud.com) is a free service to calculate sales tax
and generate tax reports. The `tax_cloud` gem allows you to easily integrate
with TaxCloud's API.

### Getting Started
Create a [TaxCloud](http://www.taxcloud.com) merchant account at
http://www.taxcloud.net. Add a website to your account under
[Locations](https://taxcloud.net/account/locations). This will generate an API
ID and API Key that you will need to use the service.

[TaxCloud](http://www.taxcloud.com) also offers an optional address
verification API. To use it, you need a USPS (United States Postal Service)
Address API Username. To obtain your USPS Username:
1.  Go through the Web Tools API Portal registration process at
    https://registration.shippingapis.com/
2.  Once you have registered for your account, you will receive an email with
    your USPS username and password.
3.  Enter your USPS username in your TaxCloud initializer (see below).


### Setup
Add the gem to your Gemfile.

```ruby
gem 'tax_cloud'
```

Configure your environment. For example, create an initializer in Rails in
`config/initializers/tax_cloud.rb`.

```ruby
TaxCloud.configure do |config|
  config.api_login_id = 'your_tax_cloud_api_login_id'
  config.api_key = 'your_tax_cloud_api_key'
  config.usps_username = 'your_usps_username' # optional
  config.open_timeout = 1 # optional
  config.read_timeout = 1 # optional
end
```

The `open_timeout` and `read_timeout` options are used to specify waiting time
for the TaxCloud web service response in seconds. Default values:
`open_timeout = 2` and `read_timeout = 2`.

### Using TaxCloud
Define the destination and origin addresses using `TaxCloud::Address`.

```ruby
origin = TaxCloud::Address.new(
  address1: '162 East Avenue',
  address2: 'Third Floor',
  city: 'Norwalk',
  state: 'CT',
  zip5: '06851'
)
destination = TaxCloud::Address.new(
  address1: '3121 West Government Way',
  address2: 'Suite 2B',
  city: 'Seattle',
  state: 'WA',
  zip5: '98199'
)
```

Create your Transaction and set up your cart items

```ruby
transaction = TaxCloud::Transaction.new(
  customer_id: '1',
  cart_id: '1',
  origin: origin,
  destination: destination
)
transaction.cart_items << TaxCloud::CartItem.new(
  index: 0,
  item_id: 'SKU-100',
  tic: TaxCloud::TaxCodes::GENERAL,
  price: 10.00,
  quantity: 1
)
lookup = transaction.lookup # this will return a TaxCloud::Responses::Lookup instance
lookup.tax_amount # total tax amount
lookup.cart_items.each do |cart_item|
  cart_item.tax_amount # tax for a single item
end
```

After you've authorized and captured the transaction via your merchant
account, you should do the same with TaxCloud for maintaining accurate tax
information.

```ruby
transaction.order_id = 100
transaction.authorized_with_capture # returns "OK" or raises an error
```

Later, you may need to mark some cart items as returned. TaxCloud will ignore
any cart items that you don't include.

```ruby
transaction.order_id = 100
transaction.cart_items << TaxCloud::CartItem.new(
  index: 0,
  item_id: 'SKU-100',
  tic: TaxCloud::TaxCodes::GENERAL,
  price: 10.00,
  quantity: 1
)
transaction.returned # returns "OK" or raises an error
```

### Verifying Addresses

[TaxCloud](http://www.taxcloud.com) optionally integrates with the USPS
Address API. An address can be verified, which can also yield a 9-digit zip
code that helps determine a more accurate tax rate.

```ruby
address = TaxCloud::Address.new({
                                  address1: '888 6th Ave',
                                  city: 'New York',
                                  state: 'New York',
                                  zip5: '10001'
                                })

verified_address = address.verify
verified_address.zip5 # 10001
verified_address.zip4 # 3502
verified_address.zip # 10001-3502
```

### Tax Codes
[TaxCloud](http://www.taxcloud.com) maintains a list of all Taxability
Information Codes or TICs, which can be found at https://taxcloud.net/tic.

You can obtain all tax codes as well as lookup a tax code by ID.

```ruby
TaxCloud::TaxCodes.all # a hash of all codes
tax_code = TaxCloud::TaxCodes[TaxCloud::TaxCodes::DIRECT_MAIL_RELATED]
tax_code.ticid # 11000 or TaxCloud::TaxCodes::DIRECT_MAIL_RELATED
tax_code.description # "Direct-mail related"
```

Tax codes are organized in groups.

```ruby
TaxCloud::TaxCode::Groups.all # a hash of all groups
tax_code_group = TaxCloud::TaxCode::Groups[TaxCloud::TaxCode::Groups::SCHOOL_RELATED_PRODUCTS]
tax_code_group.group_id # 3 or TaxCloud::TaxCode::Groups::SCHOOL_RELATED_PRODUCTS
tax_code_group.description # School Related Products
tax_code_group.tax_codes # a hash of all codes in this group
```

Tax code constants are defined in `tax_code_constants.rb` and tax code group
constants in `tax_code_group_constants.rb`. These files can be generated by
running the following rake tasks.

```
TAXCLOUD_API_LOGIN_ID=... TAXCLOUD_API_KEY=... TAXCLOUD_USPS_USERNAME=... tax_cloud:tax_codes
TAXCLOUD_API_LOGIN_ID=... TAXCLOUD_API_KEY=... TAXCLOUD_USPS_USERNAME=... tax_cloud:tax_code_groups
```

### Tax States
TaxCloud manages a list of states in which you can calculate sales tax. The
default setup will only have SSUTA (Streamlined Sales and Use Tax Agreement)
states enabled. All other states will return $0 for tax values. To enable
other states, go to https://taxcloud.com/go/states-management/. You can find
more information about SSUTA
[here](http://www.streamlinedsalestax.org/index.php?page=About-Us).

### Running Tests
*   VCR and WebMock are used to replay requests and avoid hitting the API each
    time. To refresh the mocks, simply delete the `test/cassettes` directory.
*   Run tests.
        rake test

*   If you need to record new requests against TaxCloud, set your keys first.
        TAXCLOUD_API_LOGIN_ID=... TAXCLOUD_API_KEY=... TAXCLOUD_USPS_USERNAME=... rake test

    The mocks will filter out your configuration details.


### Bugs, fixes, etc
*   Fork.
*   Write test(s).
*   Fix.
*   Commit.
*   Submit pull request.


### License

Copyright Drew Tempelmeyer and contributors, 2011-2020.

This gem is licensed under the MIT license.


CHANGELOG.md
### 0.6.1 (Next)

*   Your contribution here.

### 0.6.0 (2023/12/15)

*   Add Residential Delivery Indicator (rdi) to `TaxCloud::Address` on successful address verification - @nononoy.

### 0.5.0 (2023/7/21)

*   Update README, CHANGELOG, and LICENSE from RDoc to Markdown -
    @brchristian.
*   Migrate from Travis CI to GitHub Actions -
    @rchekaluk.
*   Update WSDL endpoint URL as recommended by TaxCloud to
    https://api.taxcloud.com/1.0/?wsdl - @mrmarcondes.


### 0.4.0 (2020/12/7)

*   Add `open_timeout` and `read_timeout` options into configuration -
    @ka8725.
*   Switch to using BigDecimal for `tax_amount` to avoid floating point
    arithmetic errors - @paulhenrich.
*   Require ruby 2.6.0+
*   Upgrade to rubocop latest
    *   Add `TargetRubyVersion: 2.6` to rubocop config (based on
        https://www.ruby-lang.org/en/downloads/branches/ oldest "normal
        maintenance" version)
    *   Add `required_ruby_version` to gemspec
    *   addresses rubocop violations

*   Update WSDL endpoint URL as recommended by TaxCloud to
    https://asmx.taxcloud.com/1.0/?wsdl - @brchristian.


### 0.3.0 (2014/1/9)

*   #19: Support Savon 2 - @drewtempelmeyer.
*   Implemented Rubocop, Ruby style linter - @dblock.
*   Ruby 1.8.7 and 1.9.2 are no longer supported - @dblock.


### 0.2.2 (2013/4/29)

*   Relaxed thirdparty gem dependency versions - @mperham.


### 0.2.1 (2013/3/3)

*   Fixed date formatting in API requests with localized applications -
    @soulcutter.


### 0.2.0 (2012/11/26)

*   The gem is now licensed under the MIT license - @dblock.
*   Raise specialized configuration and SOAP errors - @dblock.
*   Added `TaxCloud::Client.ping` - @dblock.
*   Returning a verified `TaxCloud::Address` from `TaxCloud::Address.verify` -
    @dblock.
*   Added `TaxCloud::Address.zip` that returns a 9-digit address zip code,
    when available - @dblock.
*   Returning a `TaxCloud::Responses::Lookup` from
    `TaxCloud::Transaction.lookup` - @dblock.
*   Returning transaction state, ie. `"OK"` from all other TaxCloud
    transactions. Exceptions are raised on error - @dblock.
*   Added support for tax codes and tax code groups lookup with
    `TaxCloud::TaxCodes` and `TaxCloud::TaxCode::Groups` - @dblock.


### 0.1.5 (2012/5/9)

*   Fixed compatibility with Ruby 1.8.7 and 1.9.3, removed super from
    constructors for classes which inherit from `Object` - @gfmurphy.


### 0.1.4 (2011/10/20)

*   Upgraded rdoc - @drewtempelmeyer.
*   Fixed `.gemspec` dependency declaration for savon - @drewtempelmeyer.


### 0.1.3 (2011/9/19)

*   Added support for the "returned" request - @danielmorrison.
*   Refactoreted tests to use vcr and webmock - @danielmorrison.


### 0.1.0 (2011/8/23)

*   Initial public release - @drewtempelmeyer.



examples/lookup_example.rb
##
# An example of doing a tax lookup that hits TaxCloud.  To run:
#
# bundle exec ruby lookup_example.rb
#
# Make sure you copy the .env.example to .env an update the
# values in it before running.  I also recommend no running
# this script against a live TaxCloud site.  Run it only against
# test sites.
##

require 'dotenv/load'
require 'securerandom'

require_relative '../lib/tax_cloud'

##
# Load the TaxCloud settings from the environment file.\
#
def load_config
  TaxCloud.configure do |config|
    config.api_login_id = ENV['TAX_CLOUD_LOGIN_ID']
    config.api_key = ENV['TAX_CLOUD_API_KEY']
    config.open_timeout = 10
    config.read_timeout = 10
  end
end

##
# Display the current TaxCloud settings.
#
def display_config
  TaxCloud.configure do |config|
    puts "API Login ID: #{config.api_login_id}"
    puts "API Key: #{config.api_key}"
    puts "Open timeout: #{config.open_timeout}"
    puts "Read timeout: #{config.read_timeout}"
  end
end

##
# Create a test transaction with one cart item.
#
# Will create a new customer and cart each time using GUIDs for
# the customer and cart IDs.  This prevents collisions with
# existing customer/carts when running the script multiple times.
#
def create_test_transaction
  origin = TaxCloud::Address.new(
      :address1 => '162 East Avenue',
      :address2 => 'Third Floor',
      :city => 'Norwalk',
      :state => 'CT',
      :zip5 => '06851')

  destination = TaxCloud::Address.new(
      :address1 => '3121 West Government Way',
      :address2 => 'Suite 2B',
      :city => 'Seattle',
      :state => 'WA',
      :zip5 => '98199')

  transaction = TaxCloud::Transaction.new(
      :customer_id => SecureRandom.uuid,
      :cart_id => SecureRandom.uuid,
      :origin => origin,
      :destination => destination)

  transaction.cart_items << TaxCloud::CartItem.new(
      :index => 0,
      :item_id => 'SKU-100',
      :tic => TaxCloud::TaxCodes::GENERAL,
      :price => 10.00,
      :quantity => 1)

  transaction
end

##
# Displays information about the transaction:
#
# - Customer ID
# - Cart ID
# - Origin Address
# - Destination Address
# - Cart Items
#
# @param [TaxCloud::Transaction] transaction
#
def display_transaction(transaction)
  puts "Customer ID: #{transaction.customer_id}"
  puts "Cart ID: #{transaction.cart_id}"
  puts "Origin: #{transaction.origin.to_hash}"
  puts "Destination: #{transaction.destination.to_hash}"
  puts "Cart Items: #{transaction.cart_items.map(&:to_hash)}"
end

##
# Display the total tax and tax for each cart item.
#
# @param [TaxCloud::Responses::Lookup] lookup
def display_tax(lookup)
  puts "Total tax amount: #{lookup.tax_amount}"

  lookup.cart_items.each do |cart_item|
    puts "Tax cart index #{cart_item.cart_item_index}: #{cart_item.tax_amount}"
  end
end


##
# Run an example lookup
##
puts 'Loading config information from .env file...'
load_config
puts 'Config loaded'
puts

puts 'The TaxCloud config is:'
display_config
puts

puts 'Checking config...'
TaxCloud.configuration.check!
puts 'Check passed.'
puts

puts 'Pinging TaxCloud server...'
puts TaxCloud.client.ping

puts 'Cart values:'
transaction = create_test_transaction
display_transaction(transaction)
puts

puts 'Tax lookup:'
lookup = transaction.lookup
display_tax(lookup)
puts